perf/rdm_mbw_mr.c

```diff
@@ -0,0 +1,650 @@
+			}
+
+			for(j = 0; j < window_size; j++) {
+				fi_rc = fi_recv(ep, r_buf, size, NULL,
+						fi_addrs[target], NULL);
+				assert(!fi_rc);
+			}
+
+			wait_for_data_completion(rcq, window_size);
+			fi_rc = fi_send(ep, s_buf, 4, NULL,
+					fi_addrs[target], NULL);
+			assert(!fi_rc);
+			wait_for_data_completion(scq, 1);
+		}
+	} else {
+		FT_Barrier();
+	}
+
+	ts = malloc(sizeof(t) * numprocs);
+	assert(ts);
+	FT_Allgather(&t, sizeof(t), ts);
+	if (!myid) {
+		for (i = 0; i < numprocs; i++) {
+			if (ts[i] > maxtime) {
+				maxtime = ts[i];
+			}
+		}
+	}
+	free(ts);
+
+	if(rank == 0) {
+		double tmp = num_pairs * size / 1e6;
+
+		tmp = tmp * loop * window_size;
+		bw = tmp / maxtime;
+
+		return bw;
+	}
+
+	return 0;
+}
+
+
+int main(int argc, char *argv[])
+{
+	int op, ret;
+	buf_desc_t lbuf_desc;
+
+	char *s_buf, *r_buf;
+	int align_size;
+	int pairs, print_rate;
+	int window_varied;
+	int c;
+	int curr_size;
+
+	FT_Init(&argc, &argv);
+	FT_Rank(&myid);
+	FT_Job_size(&numprocs);
+
+	/* default values */
+	pairs            = numprocs / 2;
+	window_size      = DEFAULT_WINDOW;
+	window_varied    = 0;
+	print_rate       = 1;
+
+	hints = fi_allocinfo();
+	if (!hints)
+		return -1;
+
+	while ((op = getopt(argc, argv, "f:hp:w:vr:")) != -1) {
+		switch (op) {
+		case 'f':
+			hints->fabric_attr->prov_name = strdup(optarg);
+			break;
+		case 'p':
+			pairs = atoi(optarg);
+			if(pairs > (numprocs / 2)) {
+				if(!myid) print_usage();
+				return -1;
+			}
+			break;
+		case 'w':
+			window_size = atoi(optarg);
+			break;
+		case 'v':
+			window_varied = 1;
+			break;
+		case 'r':
+			print_rate = atoi(optarg);
+			if(0 != print_rate && 1 != print_rate) {
+				if(!myid) print_usage();
+				return -1;
+			}
+			break;
+		case '?':
+		case 'h':
+			print_usage();
+			return -1;
+		}
+	}
+
+	hints->ep_attr->type	= FI_EP_RDM;
+	hints->caps		= FI_MSG | FI_RMA;
+	hints->mode		= ~0;
+
+	if(numprocs < 2) {
+		if(!myid) {
+			fprintf(stderr, "This test requires at least two processes\n");
+		}
+		FT_Finalize();
+		return -1;
+	}
+
+	/* Fabric initialization */
+	ret = init_fabric();
+	if(ret) {
+		fprintf(stderr, "Problem in fabric initialization\n");
+		return ret;
+	}
+
+	ret = init_av();
+	if(ret) {
+		fprintf(stderr, "Problem in AV initialization\n");
+		return ret;
+	}
+
+	/* Data initialization */
+	align_size = getpagesize();
+	assert(align_size <= MAX_ALIGNMENT);
+
+	s_buf = (char *) (((unsigned long) s_buf_original + (align_size - 1)) /
+				align_size * align_size);
+	r_buf = (char *) (((unsigned long) r_buf_original + (align_size - 1)) /
+				align_size * align_size);
+
+	ret = fi_mr_reg(dom, r_buf, MY_BUF_SIZE, 0, 0, 0, 0, &mr, NULL);
+	if (ret) {
+		FT_PRINTERR("fi_mr_reg", ret);
+		return -1;
+	}
+
+	lbuf_desc.addr = (uint64_t)r_buf;
+	lbuf_desc.key = fi_mr_key(mr);
+
+	rbuf_descs = (buf_desc_t *)malloc(numprocs * sizeof(buf_desc_t));
+
+	/* Distribute memory keys */
+	FT_Allgather(&lbuf_desc, sizeof(lbuf_desc), rbuf_descs);
+
```

Comment:
Looks like rbuf_descs is not needed for this benchmark. I think we can remove the FT_Allgather also.